<a href="https://github.com/ITouray/CS-448/blob/main/Assignment%204%20-%20Judging%20Flowers.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [ ]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import datasets
from sklearn import svm
from sklearn import metrics
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.datasets import make_classification
from sklearn.neural_network import MLPClassifier
from IPython.display import display, Math, Latex
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

# Part 1: Load the dataset


In [ ]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()

In [ ]:
# Output the first 15 rows of the data
df = pd.DataFrame(iris.data)
df.head(15)

0    1    2    3
0   5.1  3.5  1.4  0.2
1   4.9  3.0  1.4  0.2
2   4.7  3.2  1.3  0.2
3   4.6  3.1  1.5  0.2
4   5.0  3.6  1.4  0.2
5   5.4  3.9  1.7  0.4
6   4.6  3.4  1.4  0.3
7   5.0  3.4  1.5  0.2
8   4.4  2.9  1.4  0.2
9   4.9  3.1  1.5  0.1
10  5.4  3.7  1.5  0.2
11  4.8  3.4  1.6  0.2
12  4.8  3.0  1.4  0.1
13  4.3  3.0  1.1  0.1
14  5.8  4.0  1.2  0.2

In [ ]:
df.columns = iris.feature_names
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [ ]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [ ]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [ ]:
iris.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [ ]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

## About the dataset

Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?



#### The data shows that there are no null values. 150 observations ranging from 0-49. The features are the sepal's length and width and petal's length and width. The labels are setosa[0], versicolor[1], virginica[2].

# Part 2: Split the dataset into train and test


In [ ]:
# Take the dataset and split it into our features (X) and label (y)
x = iris.data
y = iris.target

x.shape


(150, 4)

In [ ]:
y.shape

(150,)

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = .1, random_state = 43)

In [ ]:
x_train.shape , x_test.shape , y_train.shape , y_test.shape

((135, 4), (15, 4), (135,), (15,))

# Part 3: Logistic Regression


In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(x_train, y_train)


LogisticRegression()

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred = model.predict(x_test)
y_pred

array([0, 0, 2, 1, 2, 0, 2, 1, 1, 1, 0, 1, 2, 0, 1])

In [ ]:
# iii. Report on the score for Logistic regression model, what does the score measure?
model.score(x_test, y_test)

1.0

#### A score of 1.0 shows a perfect model. It measures the accuracy of the the model. Also shows that all of the data is correctly classified.


In [ ]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
model.coef_ 

array([[-0.44068746,  0.91539817, -2.44583083, -1.0523871 ],
       [ 0.47785526, -0.26873746, -0.21782263, -0.86830965],
       [-0.0371678 , -0.64666071,  2.66365346,  1.92069675]])

In [ ]:
model.intercept_

array([  9.88651812,   2.28780948, -12.1743276 ])

# Part 4: Support Vector Machine


In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
classifier = svm.SVC(probability = True)
classifier.fit(x_train, y_train)

SVC(probability=True)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class

#y_predict = classifier.predict(x_test)
#y_predict

y_predict = classifier.predict_proba(x_test)
y_predict

array([[0.95765532, 0.03113267, 0.01121202],
       [0.97414513, 0.01698444, 0.00887043],
       [0.00973789, 0.50669165, 0.48357046],
       [0.00838264, 0.98822979, 0.00338758],
       [0.01159366, 0.00108964, 0.98731671],
       [0.96794503, 0.02229806, 0.00975691],
       [0.01051947, 0.09434651, 0.89513402],
       [0.00808731, 0.9875812 , 0.0043315 ],
       [0.01490199, 0.98130603, 0.00379198],
       [0.0565148 , 0.93088714, 0.01259806],
       [0.97349019, 0.01692056, 0.00958925],
       [0.0069433 , 0.8179992 , 0.17505749],
       [0.01275244, 0.00125821, 0.98598935],
       [0.96615366, 0.02353237, 0.01031396],
       [0.00543648, 0.93132564, 0.06323788]])

In [ ]:
# iii. Report on the score for the SVM, what does the score measure?
classifier.score(x_test, y_test)

0.9333333333333333

#### My SVM model has a high score of 93% this means that only 7% of the data is missclassified using SVM.

# Part 5: Neural Network


In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
MLP = MLPClassifier(max_iter = 500, activation = 'relu')
MLP

MLPClassifier(max_iter=500)

In [ ]:
MLP.fit(x_train, y_train)

/Users/idatouray/opt/anaconda3/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(max_iter=500)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
predictions = MLP.predict(x_test)
predictions

array([0, 0, 2, 1, 2, 0, 2, 1, 1, 1, 0, 1, 2, 0, 1])

In [ ]:
print(classification_report (y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         5
           1       1.00      1.00      1.00         6
           2       1.00      1.00      1.00         4

    accuracy                           1.00        15
   macro avg       1.00      1.00      1.00        15
weighted avg       1.00      1.00      1.00        15



In [ ]:
# iii. Report on the score for the Neural Network, what does the score measure?
print(confusion_matrix(y_test, predictions))

[[5 0 0]
 [0 6 0]
 [0 0 4]]


In [ ]:

MLP.score(x_test, y_test)

1.0

#### The score shows a perfect matrix and there are no misclassified data.

In [ ]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

#### highest score was 1.0

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

neigh = KNeighborsClassifier(n_neighbors = 5)
neigh.fit(x_train, y_train)

KNeighborsClassifier()

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class

yp = neigh.predict(x_test)
yp

array([0, 0, 2, 1, 2, 0, 2, 1, 1, 1, 0, 1, 2, 0, 1])

In [ ]:
# iii. Report on the score for kNN, what does the score measure?
print(confusion_matrix(y_test, yp))

[[5 0 0]
 [0 6 0]
 [0 0 4]]


#### The score using a confusion matrix shows no misclassified data. My score using confusion matrix shows a perfect diagonal matrix.

# Part 7: Conclusions and takeaways


In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

All of my models have a perfect score except SVM. The iris dataset is small, and that data were already well classified. No much surpeises, I sort of expected my models to have different scores with not much differences between them.